File: Shared/Entities/Deck.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using MobileCardGames.Shared.Enums;

namespace MobileCardGames.Shared.Entities
{
	public class Deck
	{
		public Deck()
		{
			cards = new Stack<PlayingCard>(DeckSize);
			Reset();
		}

		private const int DeckSize = 52;

		private readonly Stack<PlayingCard> cards;

		/// <summary>
		/// Number of cards currently in the deck
		/// </summary>
		public int Count => cards.Count;

		/// <summary>
		/// Empty current deck and fill it with new cards
		/// </summary>
		public void Reset()
		{
			cards.Clear();
			foreach (PlayingCardSuit suit in Enum.GetValues(typeof(PlayingCardSuit)))
			{
				foreach (PlayingCardValue value in Enum.GetValues(typeof(PlayingCardValue)))
				{
					cards.Push(new PlayingCard(value, suit));
				}
			}
		}

		/// <summary>
		/// Draw a card from the deck
		/// </summary>
		public PlayingCard? Draw()
		{
			return cards.Any()
				? cards.Pop()
				: default;
		}

		/// <summary>
		/// Draw cards until the deck is empty
		/// </summary>
		private IEnumerable<PlayingCard> DrawAll()
		{
			while (cards.Any())
			{
				yield return cards.Pop();
			}
		}

		/// <summary>
		/// Shuffle remaining cards in deck
		/// </summary>
		public void Shuffle()
		{
			var random = new Random();
			var allCards = DrawAll().ToList();

			while (allCards.Any())
			{
				var i = random.Next(allCards.Count - 1);
				cards.Push(allCards[i]);
				allCards.RemoveAt(i);
			}
		}
	}
}

File: game/src/entities/deck.rs
use strum::IntoEnumIterator;
use crate::entities::PlayingCard;
use crate::enums::{PlayingCardSuit, PlayingCardValue};

pub struct Deck {
	cards: Vec<PlayingCard>,
}

impl Deck {
	/// Get a 52 card deck of playing cards
	pub fn new() -> Self {
		let mut deck = Self {
			cards: Vec::with_capacity(52_usize),
		};

		deck.reset();
		deck
	}

	// Put all cards back in the deck, in order
	pub fn reset(&mut self) {
		self.cards.clear();

		for suit in PlayingCardSuit::iter() {
			for value in PlayingCardValue::iter() {
				self.cards.push(PlayingCard::new(value, suit));
			}
		}
	}

	/// Shuffle deck
	pub fn shuffle(&mut self) {
		fastrand::shuffle(&mut self.cards);
	}

	/// Draw a card, removing it from the deck.
	/// Returns [`None`] if deck is empty
	pub fn draw(&mut self) -> Option<PlayingCard> {
		self.cards.pop()
	}

	/// Number of cards currently in the deck
	pub fn len(&self) -> usize {
		self.cards.len()
	}
}